MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift

```diff
@@ -0,0 +1,143 @@
+    ///   - expiryInterval: The expiry interval used in the cache. Default value is 300 seconds (5 minutes).
+    public init(_ dataSource : D, expiryInterval: TimeInterval = 300) {
+        self.dataSource = dataSource
+        self.expiryInterval = expiryInterval
+    }
+    
+    private func isValid(_ date : Date) -> Bool {
+        return date.timeIntervalSinceNow <= expiryInterval
+    }
+    
+    public func get(_ query: Query) -> Future<T> {
+        switch query {
+        case let query as KeyQuery:
+            if let cached = objects[query.key], isValid(cached.date) {
+                return Future(cached.object)
+            }
+            return dataSource.get(query).filter { object in
+                self.objects[query.key] = (object, Date())
+                }.recover { error in
+                    self.objects[query.key] = nil
+                    return Future(error)
+                }
+        default:
+            print("TimedCacheDataSource can't cache the result of the \(type(of: dataSource)).get call because \(type(of: query)) doesn't conform to KeyQuery.")
+            return dataSource.get(query)
+        }
+    }
+    
+    public func getAll(_ query: Query) -> Future<[T]> {
+        switch query {
+        case let query as KeyQuery:
+            if let cached = arrays[query.key], isValid(cached.date) {
+                return Future(cached.array)
+            }
+            return dataSource.getAll(query).filter { array in
+                self.arrays[query.key] = (array, Date())
+                }.recover { error in
+                    self.arrays[query.key] = nil
+                    return Future(error)
+            }
+        default:
+            print("TimedCacheDataSource can't cache the result of the \(type(of: dataSource)).getAll call because \(type(of: query)) doesn't conform to KeyQuery.")
+            return dataSource.getAll(query)
+        }
+    }
+    
+    @discardableResult
+    public func put(_ value: T? = nil, in query: Query) -> Future<T> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.put(value, in: query).filter { object in
+                    self.objects[query.key] = (object, Date())
+                }.recover { error in
+                    return Future(error)
+                }
+        default:
+            print("TimedCacheDataSource can't cache the result of the \(type(of: dataSource)).put call because \(type(of: query)) doesn't conform to KeyQuery.")
+            return dataSource.put(value, in: query)
+        }
+    }
+    
+    @discardableResult
+    public func putAll(_ array: [T], in query: Query) -> Future<[T]> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.putAll(array, in: query).filter { array in
+                self.arrays[query.key] = (array, Date())
+                }.recover { error in
+                    return Future(error)
```

Comment:
I'm not sure what should we do in this case:

a) Should we remove any previously cached values for that key here?
b) Should we ignore it and do nothing?

In any case, the code as it is now it is doing nothing and we should remove the "recover" block unless we want to apply the option a).

Reply:
We should remove any previously cached values as we are not aware the business logic of the child datasource.

Reply:
will remove cached object on any error get/put/Delete.